Make table cache updates exception safe

bhalevy opened the pull request:
Currently, the row_cache external_updater:s used by the table code aren't exception safe.
That may lead to leaving the table in an invalid internal state if an the external_updater::execute throws
in the middle of cache update or invalidate.  This was one of the reasons for d4e7783188508d7d08635a2cedfcb895e50cab9e (See #10297)

This series fixes several exception safety issue and enforces exception safety of the external_updater::execute method when called form row_cache::do_update.

In general, the convention is that the `prepare` phase only updates temporary state that is cloned from the current state and modified. Therefore, `prepare` is exception safe. The `execute` methods were changed to apply the prepared temporary state and are non-throwing.  Any exception in `execute` will cause abort().

In addition to preparing temporary copies of the modified sstable sets and backlog tracker in `prepare()` and applying them in non-throwing code in `execute()`, a new sstable set implementation was added: `table_sstable_set`.
This sstable set dynamically uses the table's compaction groups to traverse all the table's sstable sets without copying all the table's sstable sets - which required table::refresh_compound_sstable_set to recalculate the overall compound_sstable_set after each update (even of a single sstable_set in a single compaction_group).

table_sstable_set takes advantage of the the compaction groups' disjointness from each other.  It creates a compound_sstable_set for each compaction group it's using to provide atomicity at the single compaction_group level but it doesn't guarantee atomicity across compaction_groups.

Fixes #10297
Refs #13937

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1482/

michoecho reviewed (changes requested):
Looks good to me overall.

I left a few comments: one about a lifetime bug (confirmed by CI), one about a (benign) case of lambda coroutine fiasco, a few typo nitpicks, and a few questions, mostly about the test in row_cache_test.

bhalevy commented on test/boost/row_cache_test.cc:
```diff
@@ -1191,7 +1229,7 @@ SEASTAR_TEST_CASE(test_update_failure) {
         }
 
-        auto ev = tracker.region().evictor();
+        auto& ev = tracker.region().evictor();
```
@tgrabiec can you please shed some light on that?

bhalevy replied:
removed the `auto&` change in v2 (and kept the existing code)
also fixed the `evicitons_left` typo in v2

bhalevy commented on test/boost/row_cache_test.cc:
```diff
@@ -1178,7 +1192,31 @@ SEASTAR_TEST_CASE(test_update_failure) {
+        };
+        auto updater = row_cache::external_updater(std::make_unique<updater_impl>(test_mode));
+
+        // fill transient memory, leaving one block free
```
block tefers to allocator's unit of allocation.
when we hit bad_alloc, the assumption here is that the last remaining block is insufficient for the requested allocation, therefore it remains free when bad_alloc is thrown.
But I'm not sure, in retrospect, that it's worth a comment.

michoecho replied:
> block tefers to allocator's unit of allocation.

I don't think this term is used anywhere else, so it should be changed or explained within the comment.

Also, the allocator doesn't have such a thing as a "unit" of allocation. It's a buddy allocator (plus a collection of slab allocator for small-sized allocations, which take their memory pools from the buddy allocator).

I don't see any guarantee that there is *any* free memory in the allocator after this bad_alloc.

bhalevy replied:
simplified the comment in v2

bhalevy commented on test/boost/row_cache_test.cc:
```diff
@@ -1178,7 +1192,31 @@ SEASTAR_TEST_CASE(test_update_failure) {
+                }
+                allocate();
+            }
+            virtual void execute() noexcept override {
```
will fix. it's from an earlier version of the series that tolerated exceptions from execute.

bhalevy replied:
fixed commit message in v2

bhalevy commented on sstables/sstable_set.cc:
```diff
@@ -363,6 +368,61 @@ void partitioned_sstable_set::insert(shared_sstable sst) {
+
+    auto undo_all_runs_inserts = defer([&] () {
+        for (const auto& sst : ssts) {
+            _all_runs[sst->run_identifier()].erase(sst);
```
valid point. we should search for it first indeed.

bhalevy replied:
This code doesn't  exist in v2.  A simple insert loop (and erase rollback loop on exception) is used instead.
Sent https://github.com/scylladb/scylladb/pull/14054 to deal with the preexisting issue in `erase()`

bhalevy commented on sstables/sstable_set.cc:
```diff
@@ -505,6 +565,19 @@ void time_series_sstable_set::insert(shared_sstable sst) {
   }
 }
 
+void time_series_sstable_set::insert(const std::vector<shared_sstable>& ssts) {
+    try {
+        for (const auto& sst : ssts) {
+            insert(sst);
+        }
+    } catch (...) {
+        for (const auto& sst : ssts) {
+            erase(sst);
+        }
+        throw;
+    }
+}
+
```
because each sst can go into a different time window, so the simple loop here at the top of the hierarchy is trivial.
We can sort the sstables by time window and then insert them in batches by using ranges from the sorted vector, but I"m not sure it's worth it.
@raphaelsc WDYT?

michoecho replied:
I meant, why can't a similar loop be used for `partitioned_sstable_set`?

bhalevy replied:
Well, it can. It'd just be slightly less efficient to insert the sstables one at a time.

bhalevy replied:
You know what, since this isn't this gist of this PR, I'll keep it simple and less risky and can do the vector sstable insert optimization as a followup patch.

bhalevy replied:
in v2 this function is the default implementation and it's used both for `time_series_sstable_set` and `partitioned_sstable_set`.

bhalevy commented on replica/compaction_group.hh:
```diff
@@ -14,6 +14,7 @@
 #include "compaction/compaction_strategy_state.hh"
 #include "sstables/sstable_set.hh"
 #include "compaction/compaction_fwd.hh"
+#include "row_cache.hh"
```
probably survived a rebase that moved code around. I'll remove it if it's unneeded.

bhalevy replied:
gone in v2

bhalevy commented on replica/sstables_adder.hh:
```diff
@@ -0,0 +1,41 @@
+
+class compaction_group_sstables_adder : public row_cache::external_updater_impl {
+protected:
+    compaction_group& cg;
```
Yeah, they were incepted as structs and then changed to class.  Will rename.

bhalevy replied:
fixed in v2

bhalevy commented on replica/table.cc:
```diff
@@ -605,17 +635,11 @@ void table::update_stats_for_new_sstable(const sstables::shared_sstable& sst) no
-        update_stats_for_new_sstable(sst);
-    }), dht::partition_range::make({sst->get_first_decorated_key(), true}, {sst->get_last_decorated_key(), true}));
+    auto& cg = compaction_group_for_sstable(sst);
+    auto sstables = std::vector<sstables::shared_sstable>({sst});
```
it's quite common pattern in our codebase to use initializer lists around single objects to construct a vector of size 1.
I'm ok with changing it here

michoecho replied:
Hmm, I think I forgot for a moment how the brace syntax works. I retract this nitpick, please ignore it.

avikivity replied:
btw initializer lists are quite bad.

1. they are const and so `std::vector({some_function()})` copies even though we have a prvalue (or whatever a temporary is called today in C++)
2. there's a bad bug with gcc + coroutines + initializer lists.

bhalevy commented on compaction/compaction_manager.cc:
```diff
@@ -1882,13 +1882,10 @@ void compaction_manager::unplug_system_keyspace() noexcept {
 }
 
 double compaction_backlog_tracker::backlog() const {
-    return disabled() ? compaction_controller::disable_backlog : _impl->backlog(_ongoing_writes, _ongoing_compactions);
```
hmm, interesting. maybe I stepped on the tail of a dragon inadvertently, and it needs to be tamed.
We need to understand why the tracker is accessed after it's cleared then. That might indicate a higher-level lifetime management problem that was lurking there.
Cc @raphaelsc

michoecho replied:
The compaction_backlog_manager is calling backlog() periodically (on a timer) on all registered trackers. It seems trackers are only unregistered on destruction, so during cleanup they are still registered, and can be accessed from the timer. Before this patch, if they are accessed in this state, `disabled()` will stop the `_impl->backlog` call, now it just happens and segfaults.

Yeah, there is something wrong with the lifetimes here. Why were the trackers disabled and not unregistered in the first place?

bhalevy replied:
I'll look into this. If they are disabled right before being destroyed we should probably unregister them first and then continue to diaable/clear and destroy.

bhalevy replied:
Fixed in v2 (compaction_backlog_tracker::clear_gently() in https://github.com/scylladb/scylladb/pull/14043/commits/5a30a0173040a0b80e4b9d07b8aa17a97c64bf82)

raphaelsc replied:
yes, I don't think it's a problem to remove it as long as compaciton_manager::remove() works as intended. once the per-table state is removed, backlog tracker is destroyed meaning it's unregistered from the list.

bhalevy commented on row_cache.cc:
```diff
@@ -1343,7 +1343,10 @@ std::ostream& operator<<(std::ostream& out, row_cache& rc) {
 future<> row_cache::do_update(row_cache::external_updater eu, row_cache::internal_updater iu) noexcept {
     auto permit = co_await get_units(_update_sem, 1);
-    co_await eu.prepare();
+    co_await eu.prepare().handle_exception([&eu] (std::exception_ptr ex) -> future<> {
```
Maybe this logic is complicated enough to refactored to a external_updater run() methid that will do something like:
```c++
    return prepare().then_wrapped([this] (future<> f) {
        if (!f.failed()) {
            std::invoke([this] () noexecpt {
                execute();
            }
        )
        return std::move(f);
    }).finally([this] {
        return cleanup();
    });
```

bhalevy replied:
or, as a coroutine:
```c++
    std::exception_ptr ex;
    try {
        co_await prepare();
    } catch (...) {
          ex = std::current_exception();
    }
    if (!ex) {
        std::invoke([this] () noexcept {
            execute();
        }
    }
    co_await cleanup();
    if (ex) {
        co_await coroutine::return_exception_ptr(std::move(ex));
    }
```

bhalevy replied:
In v2: used coroutine::as_future in do_update instead of `.handle_exception`

bhalevy commented:
In v2 ([2dce537](https://github.com/scylladb/scylladb/commit/2dce537cf94c1e43abb1ba34464a5df910410afe)):
- test: row_cache_test: fix typo in `evicitons_left`
- test: row_cache_test: extend test_update_failure
  - updated commit message
  - clarified comment about memory allocation
  - removed spurious change: `auto ev -> auto& ev = tracker.region().evictor();`
  - clarified comment about exception safety
- partitioned_sstable_set: erase empty runs
  - Fixes #14052
- sstable_set: provide insert(vector<shared_sstable>) methods
  - provide a default `sstable_set_impl::insert(const std::vector<shared_sstable>&)` implementation, overridable by implementations (for throwing an error in compound_sstable_set and for further optimization)
- compaction_group: introduce compaction_group_sstables_adder
  - removed `#include` leftover
  - renamed protected member to start with `_`
  - Clarified comment about strong exception safety
- table: introduce table_sstables_adder
  - ditto
- row_cache: coroutinize do_update, row_cache: external_updater: add cleanup method
  - used coroutine::as_future in row_cache::do_update to stick to coroutine primitives
- compaction_backlog_tracker: add clone_async, clear_gently, and swap
  - unregister the tracker in clear_gently (before possibly preempting after clearing _impl) (this caused the CI failures in debug mode where null _impl was dereferenced in `compaction_backlog_tracker::backlog()` when the `adjust()` timer hit during `clear_gently()` and before the tracker was destroyed in `compaction_manager::remove(table_state& t)`
  - fixed typos in comments

bhalevy commented on test/boost/row_cache_test.cc:
```diff
@@ -1203,10 +1241,17 @@ SEASTAR_TEST_CASE(test_update_failure) {
         }
+        testlog.info("update with failure mode={} failed={} evicitons_left={}", to_string(test_mode), failed, evicitons_left);
+        // Failure in the external_updater::{prepare,execute} phases
+        // should propagte to the caller, and be exception safe,
```
fixed in v2

bhalevy commented on test/boost/row_cache_test.cc:
```diff
@@ -1191,7 +1229,7 @@ SEASTAR_TEST_CASE(test_update_failure) {
 
-        auto ev = tracker.region().evictor();
+        auto& ev = tracker.region().evictor();
         int evicitons_left = 10;
```
fixed in v2

bhalevy commented on replica/sstables_adder.hh:
```diff
@@ -0,0 +1,41 @@
+    lw_shared_ptr<sstables::sstable_set> new_sstable_set;
+public:
+    compaction_group_sstables_adder(compaction_group& cg, std::vector<sstables::shared_sstable> sstables, is_main main);
+    // Exception safe
```
fixed in v2

bhalevy commented on compaction/compaction_backlog_manager.hh:
```diff
@@ -71,6 +77,14 @@ public:
     compaction_backlog_tracker(const compaction_backlog_tracker&) = delete;
     ~compaction_backlog_tracker();
 
+    // Clones the tracker and return an unregistered, mutable copy of it.
```
fixed in v2

bhalevy commented on compaction/compaction_backlog_manager.hh:
```diff
@@ -71,6 +77,14 @@ public:
+    // Clear the tracker gently
+    future<> clear_gently() noexcept;
+    // Swap the tracker content with another one (typically a cloned one that was changed)
+    // Tracker registration of both objects is uneffected.
```
fixed in v2

avikivity commented:
Can we add tests that demonstrate the problems and the fixes? Perhaps using error injection.

avikivity commented on row_cache.cc:
```diff
@@ -1342,8 +1342,11 @@ future<> row_cache::do_update(row_cache::external_updater eu, row_cache::interna
         [&] () noexcept {
+            // Any error from execute is considered fatal
+            // to enforce exception safety.
+            eu.execute();
```
It would be nicer to catch the error, explain what's going on, then abort.

bhalevy replied:
ok, I actually had a version that did exactly that and I changed my mind to lean towards brevity.

avikivity commented on test/boost/row_cache_test.cc:
```diff
@@ -1179,7 +1193,31 @@ SEASTAR_TEST_CASE(test_update_failure) {
+            void allocate() {
+                while (true) {
+                    auto x = bytes(bytes::initialized_later(), 4 * 1024);
+                    memory.emplace_back(std::move(x));
```
How long does this take to fail?

Does it work without the seastar memory allocator? Without bringing the machine down? I guess it won't run in debug mode.

bhalevy replied:
I didn't measure but it doesn't seem long when run interactively.
And no, it doesn't run in debug mode since it's under `#ifndef SEASTAR_DEFAULT_ALLOCATOR`

avikivity commented on sstables/sstable_set.cc:
```diff
@@ -364,7 +364,12 @@ void partitioned_sstable_set::insert(shared_sstable sst) {
+        if (it->second.empty()) {
+            _all_runs.erase(it);
+        }
+    }
```
Should this be promoted into a separate pull request? It doesn't look related.

bhalevy replied:
It is.  See https://github.com/scylladb/scylladb/pull/14054
I dequeued this commit from the series in the next version.

avikivity commented on sstables/sstable_set.cc:
```diff
@@ -177,6 +177,26 @@ sstable_set::insert(shared_sstable sst) {
+    } catch (...) {
+        for (const auto& sst : ssts) {
+            erase(sst);
+        }
```
It's more correct to erase only the ones that were successfully inserted.

bhalevy replied:
Agreed.  We can maintain an undo vector for each successfully inserted sstable and erase only those.

michoecho replied:
I also thought about this during my review, but I didn't comment on it because pre-existing single-sstable `insert` also can erase an sstable that it didn't insert.

This implementation is OK as long as its forbidden to attempt an insert of an sstable that was already inserted, which I guess is true. But in general it's weird for something called a "set" to break down when the same element is inserted twice.

avikivity replied:
We never insert the same element twice (that I know of). We should consider erroring out in that case, because something went wrong, and to rename the container to satisfy the pedants.

avikivity commented on sstables/sstable_set.hh:
```diff
@@ -121,6 +122,7 @@ public:
         });
     }
     void insert(shared_sstable sst);
+    void insert(const std::vector<shared_sstable>& ssts);
```
std::span is nicer for such interfaces.

bhalevy replied:
makes sense.

michoecho replied:
@avikivity Has our ban on c++20 ranges been lifted yet?

avikivity replied:
std::span is not part of std::ranges.

The ban on std::ranges will be lifted when we move to clang 16, which fixed a clang bug that caused many uses of std::ranges to fail.

avikivity commented on replica/sstables_adder.hh:
```diff
@@ -0,0 +1,41 @@
+#include "sstables/shared_sstable.hh"
+#include "sstables/sstable_set.hh"
+
+using namespace seastar;
```
seastarx.hh

avikivity commented on replica/compaction_group.hh:
```diff
@@ -97,8 +98,10 @@ public:
     // Returns minimum timestamp from memtable list
     api::timestamp_type min_memtable_timestamp() const;
     // Add sstable to main set
+    // Exception safe
```
Better to state the guarantee (strong guarantee)

avikivity commented on replica/compaction_group.hh:
```diff
@@ -97,12 +97,6 @@ public:
-    void add_sstable(sstables::shared_sstable sstable);
-    // Add sstable to maintenance set
-    // Exception safe
-    void add_maintenance_sstable(sstables::shared_sstable sst);
```
This style of add-then-change-then-remove hinders review. The reviewer sees each change individually but it's hard to verify that "add" is equivalent to "remove". It results in smaller changes, but it's harder to see the overall picture.

bhalevy replied:
I can mention the eventual removal of those functions in the earlier patches to provide the context.
Otherwise, this can be done in the previous patch: `table: introduce table_sstables_adder`

bhalevy commented:
- removed "partitioned_sstable_set: erase empty runs"
  - pushed independently as https://github.com/scylladb/scylladb/pull/14054

avikivity commented on row_cache.hh:
```diff
@@ -179,6 +179,8 @@ public:
         // Any exceptions from execute() are fatal.
         virtual void execute() = 0;
+        // Cleanup must never fail.
+        virtual future<> cleanup() noexcept { return make_ready_future<>(); }
```
Please document what that cleanup must be executed even if prepare fails.

IMO, it's cleaner if the execution is up to the first failure: if prepare fails, it cleans up after itself. But it's okay this way too.

bhalevy replied:
cleanup was dequeued from v3

bhalevy commented:
> Can we add tests that demonstrate the problems and the fixes? Perhaps using error injection.

I extended row_cache_test.update_failure_tester to test `prepare()` error handling.
For the other cases, it's quite hard to show the issue with broken backlog tracker or table compound sstable set.
I'm not sure it's really worth the effort to add such intrusive error-injecting test, that will become pointless once the exception safety is achieved.

avikivity commented on compaction/compaction_manager.cc:
```diff
@@ -1962,7 +1962,7 @@ compaction_backlog_tracker::compaction_backlog_tracker(compaction_backlog_tracke
-        if (auto manager = std::exchange(_manager, x._manager)) {
+        if (auto manager = std::exchange(_manager, nullptr)) {
             manager->remove_backlog_tracker(this);
         }
```
It's not clean if the correctness of compaction_backlog_manager move assignment depends on what its caller does.

bhalevy replied:
I agree.  I don't think this block is even needed since this is called only from `compaction_manager::register_backlog_tracker` and `cs.backlog_tracker` is either unregistered after it was initially constructed in `: backlog_tracker(t.get_compaction_strategy().make_backlog_tracker())` or it's already registered with the same manager.

bhalevy replied:
replaced with "compaction_backlog_tracker: operator=: do not change tracker registration" in v3 that just removes this block and documents that the function does not affect registration.

avikivity commented on compaction/compaction_backlog_manager.hh:
```diff
@@ -71,6 +77,14 @@ public:
+    future<> clear_gently() noexcept;
+    // Swap the tracker content with another one (typically a cloned one that was changed)
+    // Tracker registration of both objects is unaffected.
+    void swap(compaction_backlog_tracker&) noexcept;
```
Why do we need async/gently variants? How large can these grow? How long does it take to copy/clear them?

bhalevy replied:
Unfortunately, in pathological cases a compaction group may have 1000's if not 10,000's sstables.
In can measure how long it takes to clone/cleanup an oversized backlog tracker.

avikivity replied:
thousands of tables is still peanuts (0.5ms / 1000 -> 0.5us per table, or ~2000 cycles at 2GHz).

tens of thousands is too much.

bhalevy replied:
I left that out in v3.
Will reconsider that as followup (along with sstable_set / backlog_tracker transactions)

bhalevy replied:
FWIW, we stumbled upon stalls in partitioned_sstable_set insertion and destroy.
See https://github.com/scylladb/scylla-enterprise/issues/2975

avikivity replied:
I guess partitioned_sstable_set is very slow. It's also O(n log n), so 2000 sstables = 22000 ops.

avikivity commented on replica/table.cc:
```diff
@@ -470,8 +471,7 @@ void compaction_group_sstables_adder::execute() {
         }
-        // FIXME: the following isn't exception safe.
-        table::add_sstables_to_backlog_tracker(_cg.get_backlog_tracker(), _sstables);
+        _cg.get_backlog_tracker().swap(*_new_backlog_tracker);
```
An alternative way to clone/add is to use software transactions. Each mutator returns an object that, when destroyed, undoes the operation.

The caller calls mutators and stores the objects in some vectors. Once an exception happens, everything gets rolled back. If an exception doesn't happen, we call cancel() on all the reverters.

This favors mutations that are smaller than the object being mutated (since we don't need to copy everything).

/cc @tgrabiec

bhalevy replied:
It's possible and seems cleaner.
Just bare in mind that we copy-modify-write the sets also to protect concurrent readers that iterate over the set:
See https://github.com/scylladb/scylladb/blob/af65d5a1e8e61565da6ef60ae7a4a8dab675e0f6/sstables/sstable_set.cc#L326

avikivity replied:
I thought we added versioning to make the copy very cheap.

bhalevy replied:
Versioning of sstable_set? Not that I'm aware of.

michoecho replied:
> I thought we added versioning to make the copy very cheap.

We tried, but ultimately we didn't. It was merged to master (31909515b37a096e62cbc23b63e4e3e4e574bfe7), then promptly reverted (5f4bf183874a3ff6492d85d27d5371f8cc428678), then there was another attempt (https://github.com/scylladb/scylladb/pull/8242), but it lost a merge race with `compound_sstable_set`, and it seems the project was abandoned at that point.

michoecho replied:
@wmitros Why was versioned sstable set eventually abandoned?

raphaelsc replied:
IIRC, the main reason for versioned sstable set was to address stalls, the stalls were (partially?) fixed, and we abandoned it due to cost benefit not being satisfactory. we can revisit it, and see if we can still benefit from it. with tablets, the value is even lower?!

bhalevy replied:
In v3 I kept the clone-modify-set logic.
We can consider something else as a follow-up

avikivity commented on sstables/sstable_set.cc:
```diff
@@ -325,6 +325,13 @@ std::unique_ptr<sstable_set_impl> partitioned_sstable_set::clone() const {
+    co_await utils::clear_gently(_leveled_sstables);
+    co_await utils::clear_gently(_all);
+    co_await utils::clear_gently(_all_runs);
+}
```
Why are these needed? How large are sstable sets?

bhalevy replied:
We didn't see any stalls in testing but I'm afraid that in real-life they could grow to 10's of thousands of sstables.
Keep in mind that with ICS with 1GB default fragment size, a 1TB table holds a 1000 sstables.
With tablets this should be more controlled for sure.

bhalevy replied:
Adding `cleanup()` and the `clear_gently` calls isn't instrumental to exception safety, so I can postpone that to a follow-up after proving why and when it's needed.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1523/

raphaelsc commented:
Will be reviewing it on Monday.

bhalevy commented:
In v3 ([0e904be](https://github.com/scylladb/scylladb/commit/0e904bea4feb746de8968d66fc0fce89750055b1)):
* Simplified series by not introducing clone_async, clear_gently, and external_updater::cleanup changes. Will be considered as follow-up.
* row_cache: abort on exteral_updater::execute errors
  - print reason for abort if execute fails
* sstable_set: provide insert(vector<shared_sstable>) methods
  - revert only successfully inserted sstables when handling exception
* table: introduce table_sstables_adder
  - delete unneccessary comments previously added to compaction_group.hh
* compaction_backlog_tracker: reset backlog manager in move-assignment
  - dropped in favor of "compaction_backlog_tracker: operator=: do not change tracker registration"
* compaction_backlog_tracker: add clone
  - instead of clone_async, clear_gently, and swap
* compaction_backlog_tracker: get rid of disable logic
  - No clear_gently
* compaction_group: set*sstables: add optional delta_disk_space_used param
  - To keep update of `cg._main.*_set_disk_space_used` incremental
* sstable_set: keep impl as shared_ptr
  - Allows updating the existing sstable sets in place by holding a shared_ptr<sstable_set_impl> while traversing the sstable set and updating the sstable_set _impl shared_ptr rather than the higher level lw_shared_ptr<sstable_set> since the latter requires recalculating the table partitioned_sstable_set (in the execute() function).
* table: update sstable sets in place when updating cache

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1532/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1536/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1537/

bhalevy commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1532/

In v4 ([daf778b](https://github.com/scylladb/scylladb/commit/daf778b346c2f99b685a5afa73533b77c3b3285e)):
* fixed build failures related to not including compaction_backlog_manager.hh from replica/database.hh

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1538/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1539/

avikivity commented on sstables/sstable_set.cc:
```diff
@@ -177,6 +177,29 @@ sstable_set::insert(shared_sstable sst) {
+    try {
+        for (const auto& sst : ssts) {
+            insert(sst);
+            undo_list.emplace_back(sst);
```
The undo vector is not needed, just the number of sstables successfully inserted.

avikivity commented on replica/sstables_adder.hh:
```diff
@@ -0,0 +1,41 @@
+/*
+ * Copyright (C) 2014-present ScyllaDB
```
2023?

avikivity commented on replica/sstables_adder.hh:
```diff
@@ -0,0 +1,41 @@
+#include <seastar/util/bool_class.hh>
+
+#include "seastarx.hh"
+#include "row_cache.hh"
```
You're adding a heavy header where it's not necessary, this breaks modularization and causes excessive recompilation.

avikivity replied:
Well, it is needed here.

avikivity replied:
The most accurate approach is to extract external_updater into a separate header and include only that. However, it's overkill in this case.

avikivity commented on replica/table.cc:
```diff
@@ -446,25 +447,48 @@ void compaction_group::backlog_tracker_adjust_charges(const std::vector<sstables
+    if (_main) {
+        _cg._main_sstables = std::move(_new_sstable_set);
+        for (const auto& sst : _sstables) {
+            _cg._main_set_disk_space_used += sst->bytes_on_disk();
```
Perhaps we should make sstable_set track this (not now).

bhalevy replied:
Yes, it's doable, although an optimization.

avikivity commented on replica/table.cc:
```diff
@@ -609,17 +640,11 @@ void table::update_stats_for_new_sstable(const sstables::shared_sstable& sst) no
+    auto sstables = std::vector<sstables::shared_sstable>({sst});
+    auto adder = row_cache::external_updater(std::make_unique<table_sstables_adder>(*this, cg, std::move(sstables), is_main(!bool(offstrategy))));
+    co_return co_await get_row_cache().invalidate(std::move(adder),
+            dht::partition_range::make({sst->get_first_decorated_key(), true}, {sst->get_last_decorated_key(), true}));
```
Note this can throw

bhalevy replied:
True, but that would happen before calling "invalidate" so it's orthogonal to the updater begin exception safe.
Why is it a problem?

avikivity commented on replica/table.cc:
```diff
@@ -644,26 +669,38 @@ table::add_sstables_and_update_cache(const std::vector<sstables::shared_sstable>
-            sources.push_back(sst->as_mutation_source());
+
+    class updater : public table_sstables_adder {
+        lw_shared_ptr<memtable> _m;
```
Can you explain why we need this tall hierarchy?

bhalevy replied:
vs. what?  open coding the `table_sstables_adder` in this local `updater`?

avikivity commented on replica/table.cc:
```diff
@@ -644,26 +669,38 @@ table::add_sstables_and_update_cache(const std::vector<sstables::shared_sstable>
+            , _m(std::move(m))
+        {
+            if (_sstables.size() == 1) {
+                _ms_opt = _sstables.front()->as_mutation_source();
```
Let's move this optimization to make_combined_mutation_source() (can be done later)

bhalevy replied:
makes sense

avikivity commented on compaction/compaction_manager.cc:
```diff
@@ -1962,9 +1962,6 @@ compaction_backlog_tracker::compaction_backlog_tracker(compaction_backlog_tracke
     if (this != &x) {
-        if (auto manager = std::exchange(_manager, x._manager)) {
-            manager->remove_backlog_tracker(this);
-        }
```
I don't know if this is correct. The manager will still refer to `this`, but `this->_manager` will point somewhere else.

We should unregister `this` from the current manager, unregister x from x._manager, adopt x._manager, and register ourselves in it.

bhalevy replied:
why? this->manager remains unchanged with this change.
x gets unregistered when destroyed, but I don't think this ever happens since we're assigning registered trackers with unregistered ones, we can even throw internal error if that's not the case.  We may want to move-construct registered trackers, but otherwise the move assignment is to replace the contents of the assigned tracker, not its registration as far as I could tell.

bhalevy replied:
In the next version I'll introduce a new function to do this: `transfer_backlog`, and delete the move-assignment operator to avoid confusion.

avikivity commented on compaction/compaction_backlog_manager.hh:
```diff
@@ -72,6 +73,9 @@ public:
     ~compaction_backlog_tracker();
 
+    // Clone the tracker and return an unregistered, mutable copy of it.
+    compaction_backlog_tracker clone() const;
```
Is this really necessary? It looks like the change has huge impact. What does it mean to clone a compaction_backlog_tracker? Will we report double the backlog?

There has to be a better way.

avikivity replied:
Furthermore, what about locking? This is done with preemption points. Can something sneak in and change things under our feet?

michoecho replied:
> What does it mean to clone a compaction_backlog_tracker? Will we report double the backlog?

Since the cloned tracker isn't registered anywhere, we won't double report anything. It's just a bag of data until it's registered.

> Furthermore, what about locking? This is done with preemption points. Can something sneak in and change things under our feet?

Good point. I'm not familiar with the sstables part of this PR, and I assumed during my review that everything here happens under `_sstable_set_mutation_sem`. But it seems this isn't true, and for example `_ongoing_writes` is modified outside of that semaphore, so it could happen concurrently with the read-modify-write of the tracker, dropping info about ongoing writes with some low probability.

bhalevy replied:
> What does it mean to clone a compaction_backlog_tracker? Will we report double the backlog?

No, we only clone the contents but not the registration.

> Furthermore, what about locking? This is done with preemption points. Can something sneak in and change things under our feet?

Cache updates are serialized with a semaphore.
This is true for any changes to the compaction group sstable set (adding, removing, or both on compaction completion)

michoecho replied:
> Cache updates are serialized with a semaphore.
This is true for any changes to the compaction group sstable set (adding, removing, or both on compaction completion)

Even for `_ongoing_writes` and `_ongoing_compactions`?

bhalevy replied:
> There has to be a better way.

Ulitimately, transactions, as you suggested can do better.  It's still could be a challenge to prevent cloning when there could be asynchronous readers of the state that don't symchronize with the update lock.

For sstable sets, at least, we can carry an auxiliary list of changes that can be in pending (uncommitted) state and so they are ignored by reads and they can be reverted as part of error handling.

Committing them should be noexcept, and once committed they would be considered as part of the set.  Thes set of mutations should be "compacted" into the sstable set main vector.
This can be achieved using a rwlock so cache update can try to acquire the write lock and apply the "mutations" onto the set. If that fails, the mutations are still committed, though not applied.
If the rwlock is taken by a read, cache update just leaves the committed changes behind and any read should detect them, acquire the write lock and apply them then in the beggining of for_each_sstable_gently, or after it's done.

I hope we can intwgrate that mechanism also with the backlog tracker.

bhalevy replied:
> > Cache updates are serialized with a semaphore.
> > This is true for any changes to the compaction group sstable set (adding, removing, or both on compaction completion)
> 
> Even for `_ongoing_writes` and `_ongoing_compactions`?

That's an excellent point.
I'll need to rethink this.

avikivity replied:
We have to avoid complexity here. These are rarely-called areas and so we won't have much coverage. I don't want to trade one rare failure mode for another.

raphaelsc replied:
Backlog tracker reflects the content in sstable set. I suggested before, and I think we should consider marrying sstable_set and backlog_tracker in the compaction namespace. But that's for later.

raphaelsc replied:
Benny, I don't understand the motivation for this patch, since recently you changed backlog_tracker::replace_sstables to provide strong exception safety. Is this patch really needed? When?

bhalevy replied:
Even though it's exception safe it might still throw, and we don't want that during the execute() call.
In retrospect, it's possible we didn't have to make it exception safe if we're doing the update to a temporary value in the prepare phase, but it's a good property to have in any case.

raphaelsc commented:
Postponing my review to Tuesday. Went through some discussions, but didn't have time to peek into changes yet.

tgrabiec commented on replica/table.cc:
```diff
@@ -475,7 +470,9 @@ const lw_shared_ptr<sstables::sstable_set>& compaction_group::main_sstables() co
-    _main_sstables = std::move(new_main_sstables);
+    // Readers hold a shared_ptr of the sstable_set_impl
+    // so it's safe to update the sstable_set
+    *_main_sstables = std::move(*new_main_sstables);
```
Doesn't it violate snapshot isolation for the sstable set?

Before the PR, when someone obtained `lw_shared_ptr<sstables::sstable_set>`, it could assume that it doesn't change.

For example:
```
partition_presence_checker
table::make_partition_presence_checker(lw_shared_ptr<sstables::sstable_set> sstables) {
```

It holds to `lw_shared_ptr<sstables::sstable_set>`, not the impl.

tgrabiec commented on row_cache.cc:
```diff
@@ -1337,11 +1342,8 @@ std::ostream& operator<<(std::ostream& out, row_cache& rc) {
-    }).then([this, &eu, &iu] (auto permit) mutable {
-      return eu.prepare().then([this, &eu, &iu, permit = std::move(permit)] () mutable {
+    auto permit = co_await get_units(_update_sem, 1);
+      co_await eu.prepare();
```
Can exception from prepare() abort the node because do_update() is marked as noexcept, in case we didn't defer the coroutine yet?

bhalevy replied:
Coroutines are noexcept by nature, turning all exceptions into an exceptional future return, so no terminate will happen

bhalevy replied:
This is now tested in row_cache_test.test_update_failure_in_prepare

michoecho commented on replica/sstables_adder.hh:
```diff
@@ -0,0 +1,41 @@
+    compaction_group_sstables_adder(compaction_group& cg, std::vector<sstables::shared_sstable> sstables, is_main main);
+    // Guarantees strong exception safety
+    virtual future<> prepare() override;
+    // FIXME: need to provide strong exception guarantees.
```
aborting on exceptions != expecting no exceptions

As far as I understand, the idea is that exceptions from execute() are technically allowed (they aren't a programming error) — they are just treated as fatal errors higher in the call chain.

bhalevy replied:
> aborting on exceptions != expecting no exceptions
> 
> As far as I understand, the idea is that exceptions from execute() are technically allowed (they aren't a programming error) — they are just treated as fatal errors higher in the call chain.

right

bhalevy commented on replica/table.cc:
```diff
@@ -446,25 +447,48 @@ void compaction_group::backlog_tracker_adjust_charges(const std::vector<sstables
-    _main_set_disk_space_used += sstable_size;
+    auto adder = compaction_group_sstables_adder(*this, {std::move(sstable)}, is_main::yes);
+    // FIXME: for now, prepare is essentially synchronous
+    (void)adder.prepare();
```
`add_sstable` is gone in a later patch so it's not worth the trouble.

bhalevy commented on replica/table.cc:
```diff
@@ -1956,13 +1962,17 @@ future<db::replay_position> table::discard_sstables(db_clock::time_point truncat
         tlogger.debug("cleaning out row cache");
     }));
     rebuild_statistics();
-    co_await coroutine::parallel_for_each(p->remove, [this, p] (pruner::removed_sstable& r) -> future<> {
```
Just to simplify the implementation.
Can do it in a separate patch if this needs justification.

bhalevy commented:
Submitted #15052 for sstable_set maintain bytes_on_disk

bhalevy commented on sstables/sstable_set.hh:
```diff
@@ -91,6 +91,7 @@ public:
     std::unique_ptr<sstable_set_impl> _impl;
     schema_ptr _schema;
+    uint64_t _bytes_on_disk;
 public:
```
We can virtualize `bytes_on_disk()` so it can be overridden by compound_sstable_set to return the sum of all underlying sstable sets.

avikivity replied:
Yes

raphaelsc commented on replica/table.cc:
```diff
@@ -1149,6 +1150,7 @@ compaction_group::update_sstable_lists_on_off_strategy_completion(sstables::comp
+            // FIXME: the following is not exception safe
             _t.refresh_compound_sstable_set();
             // Input sstables aren't not removed from backlog tracker because they come from the maintenance set.
             _cg.backlog_tracker_adjust_charges({}, _new_main);
```
interesting, if you recall, we wanted to stop maintaining a clone of sstable set in backlog tracker. and turns out we also suffer with exception unsafety as sstable set and backlog tracker are updated independently. perhaps we could implement a backlog_tracked_sstable_set later, so we kill two birds with one stone.

bhalevy replied:
But the backlog tracker is per compaction group, while the compound sstable set covers the whole table.

bhalevy replied:
Please review the updated series, in particular the last patches in the series:
- sstable_set: keep impl as shared_ptr
- sstable_set: keep sstable_set_impl_ptr in incremental selectors impl
- sstable_set: allow updating in-place
- sstable_set: compound_sstable_set: point to underlying sstable_set:s

They allow the compound_sstable_set to safely keep pointers to the underlying sstable sets, and allow updating those in-place by replacing the shared_ptr<sstable_set_impl> they keep when updating the cache.

With that, `table::refresh_compound_sstable_set` is no longer required.

bhalevy commented:
In v5 ([c29838d](https://github.com/scylladb/scylladb/commit/c29838d9e7826be6935ce88ad66023630653568d)):
- rebased on top of [5949623](https://github.com/scylladb/scylladb/commit/5949623e0d202d8a3d71dfaf30b113d1c29de20c)
- cleaned up patches
- sstable_set: allow updating in-place
- cql-pytest/test_select_from_mutation_fragments: do not share test_table
- sstable_set: compound_sstable_set: point to underlying sstable_set:s

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3118/

bhalevy commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3118/

The test_writes_caching_disabled dtest failed.
Investigating...

bhalevy commented:
In v6 ([7162c95](https://github.com/scylladb/scylladb/commit/7162c9523b14f440a676efa7e80abba49bb5f82e)):
- sstable_set: compound_sstable_set: point to underlying sstable_set:s
  - clone _sstables set in table::sstables_as_snapshot_source

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3135/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3158/

bhalevy commented:
> @bhalevy I guess cover letter is outdated. we no longer have cleanup phase right?

updated cover letter

raphaelsc commented:
@bhalevy I also noticed that some of the patches were already merged upstream. please rebase the branch against master, so I can review the remaining patches only

raphaelsc commented:
> @bhalevy I also noticed that some of the patches were already merged upstream. please rebase the branch against master, so I can review the remaining patches only

Sorry for the noise. I was confused by my local git state. Will review this series tomorrow.

bhalevy commented:
@raphaelsc ping review

raphaelsc commented on row_cache.hh:
```diff
@@ -177,8 +177,8 @@ public:
-        // also change function signature.
+        // Updaters that are exception safe should handle all exceptions in prepare.
+        // Any exceptions from execute() are fatal.
         virtual void execute() = 0;
```
isn't it better to keep the reminder to eventually make execute() noexcept? we're heading there.

bhalevy replied:
okay, it's a noble goal indeed.

raphaelsc commented on replica/database.hh:
```diff
@@ -562,10 +562,8 @@ private:
         return _config.enable_cache && _schema->caching_options().enabled();
     }
     void update_stats_for_new_sstable(const sstables::shared_sstable& sst) noexcept;
```
update_stats_for_new_sstable is now unused

raphaelsc commented on replica/database.hh:
```diff
@@ -562,10 +562,8 @@ private:
     }
     void update_stats_for_new_sstable(const sstables::shared_sstable& sst) noexcept;
+    void update_sstables_stats(ssize_t count, int64_t bytes_on_disk) noexcept;
     future<> do_add_sstable_and_update_cache(sstables::shared_sstable sst, sstables::offstrategy offstrategy);
```
it seems we now have the potential to calcualte these stats on the fly using sstables_set, therefore avoiding this boring independent tracking.

raphaelsc commented on replica/table.cc:
```diff
@@ -594,26 +561,99 @@ void table::update_stats_for_new_sstable(const sstables::shared_sstable& sst) no
+    virtual future<> prepare() override {
+        auto cur = _is_main ? _cg.main_sstables() : _cg.maintenance_sstables();
+        _new_sstable_set = make_lw_shared<sstables::sstable_set>(*cur);
+        auto inserted = _new_sstable_set->insert(_sstables);
```
Reuse table::sstable_list_builder? It yields if needed. I know it's the same approach as before. I am thinking about LCS, where set being closed may contain thousands of elements to be copied in a single shot using sstable_set::clone().

bhalevy replied:
I think it's an overkill as it's doing too much and even the new_sstable_list (which is actually an sstable_set) needs to be passed by value, so here we'll need to create a sstable_set containing _sstables with no real reason.

A later patch also introduces the use of sstable_set::clone()
so it we're worried about preemption, we can add clone_gently that would yield.

raphaelsc commented on row_cache.cc:
```diff
@@ -1383,29 +1383,28 @@ std::ostream& operator<<(std::ostream& out, row_cache& rc) {
+    if (f.failed()) {
+        clogger.warn("Failure during cache update: {}", f.get_exception());
+    }
 }
```
nice

raphaelsc commented on compaction/compaction_manager.cc:
```diff
@@ -2125,9 +2125,6 @@ compaction_backlog_tracker::compaction_backlog_tracker(compaction_backlog_tracke
     if (this != &x) {
-        if (auto manager = std::exchange(_manager, x._manager)) {
-            manager->remove_backlog_tracker(this);
-        }
```
looks very correct. deregistration should always be done on the dtor.

raphaelsc commented on compaction/compaction_manager.cc:
```diff
@@ -1819,9 +1819,6 @@ future<> compaction_manager::try_perform_cleanup(owned_ranges_ptr sorted_owned_r
     auto& cs = get_compaction_state(&t);
     co_await run_with_compaction_disabled(t, [&] () -> future<> {
         auto update_sstables_cleanup_state = [&] (const sstables::sstable_set& set) -> future<> {
-            // Hold on to the sstable set since it may be overwritten
```
"Keeping a shared impl in the sstable set
allows us to update just this set without
requiring rebuild of the whle table's sstable set."

what about full scans? which span all compaction groups. I guess I'll understand it later in the next patches.

bhalevy replied:
good point.
We need to deal with that in the compound_sstable_set::for_each_sstable_gently_until override

raphaelsc commented on sstables/sstable_set.cc:
```diff
@@ -794,6 +792,7 @@ static logging::logger irclogger("incremental_reader_selector");
     const dht::partition_range* _pr;
     tracing::trace_state_ptr _trace_state;
+    sstable_set _sstables;
     std::optional<sstable_set::incremental_selector> _selector;
```
unused

raphaelsc commented on replica/table.cc:
```diff
@@ -114,17 +114,20 @@ table::make_sstable_reader(schema_ptr s,
+        sstable_sets.emplace_back(&cg->main_sstables());
+        sstable_sets.emplace_back(&cg->maintenance_sstables());
+    }
     return sstables::make_compound_sstable_set(schema(), std::move(sstable_sets));
```
this is a good change. avoids one layer of indirection that comes with compound set. for tablet aware sstable set, we'll only probably need only a hint that sets belong to a single group. it's probably good to ask group for a vector of sets. so the knowledge of which sets are used is confined to compaction_group.

bhalevy replied:
The thing is that in this branch we update the sets in-place so even if e.g. some compaction_group maintenance sstable set is now empty, it might be populated later and we don't want to rebuild the table's compound sstable_set for that.

raphaelsc commented on sstables/sstable_set_impl.hh:
```diff
@@ -125,9 +125,16 @@ public:
+    compound_sstable_set(const compound_sstable_set&) = delete;
+    compound_sstable_set& operator=(compound_sstable_set&&) = default;
+    compound_sstable_set& operator=(const compound_sstable_set&) = delete;
 
```
this is a very interesting series, but I confess I need to come back tomorrow with a fresher mind. these are big changes. were there any lifetime management issue that worried you at any point in time? how do you guarantee immutability of sstable set in range scan? We cannot clone the set on every read, but rather we prefer to do a copy on write (which frequency is memtable flush). how does incremental_reader_selector keep a immutable version of sstable set without being cloned if table can now update the set in place?

bhalevy replied:
> this is a very interesting series, but I confess I need to come back tomorrow with a fresher mind. these are big changes. were there any lifetime management issue that worried you at any point in time? how do you guarantee immutability of sstable set in range scan?

That needs to be fixed in the next revision.

> We cannot clone the set on every read, but rather we prefer to do a copy on write (which frequency is memtable flush). how does incremental_reader_selector keep a immutable version of sstable set without being cloned if table can now update the set in place?

We need to keep a shared pointer of the sstable_set_impl instead of the sstable_set, since updating the sstable_set replaces the shared point it holds, so the previous shared pointer is immutable and holds a snapshot of the sstable_set contents.

bhalevy commented:
In v7 ([9977ccb](https://github.com/scylladb/scylladb/commit/9977ccb917ff0c34787a95ae7758393f80574a1b)):
- rebased
- row_cache: abort on exteral_updater::execute errors
  - kept FIXME regarding external_updater_impl::execute
- sstable_set: keep impl as shared_ptr
  - take a "snapshot" of the underlying sstable_set_impl_ptr:s in compound_sstable_set for_each_sstable_gently
- sstable_set: add clone_gently
- sstable_set: allow updating in-place
  - use clone_gently

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3310/

bhalevy commented:
> @tgrabiec please re-review.

ping @tgrabiec

raphaelsc commented on sstables/sstable_set.cc:
```diff
@@ -182,6 +182,34 @@ sstable_set::erase(shared_sstable sst) {
+        throw;
+    }
+    return inserted;
+}
```
we don't need it at sstable_set_impl. this is only a helper which reuses existing methods. for compound set, the first insertion attempt will trigger the exception.

bhalevy replied:
Good point.  Moves to sstable_set in the next rev.

raphaelsc commented on replica/table.cc:
```diff
@@ -693,32 +660,99 @@ future<sstables::sstable_list> table::take_storage_snapshot(dht::token_range tr)
+        auto inserted = _new_sstable_set->insert(_sstables);
+        if (inserted.size() != _sstables.size()) {
+            tlogger.warn("table_sstables_adder: some sstables are already in the table: candidates={} inserted={}", _sstables, inserted);
+            _sstables = std::move(inserted);
```
we have known sources of new sstables, the ones in this path are:

repair
sstable loading (on refresh on boot)
memtable

if tried to add sstable A and it's already there, it means that something have gone terribly wrong. a given sstable should never be added twice to table's sstable set. it's added once and removed once. so it seems to me on_internal_error() is better here.

bhalevy replied:
okay, I'll change that in the next version.

raphaelsc commented on compaction/compaction_manager.cc:
```diff
@@ -2128,9 +2128,6 @@ compaction_backlog_tracker::compaction_backlog_tracker(compaction_backlog_tracke
 compaction_backlog_tracker&
 compaction_backlog_tracker::operator=(compaction_backlog_tracker&& x) noexcept {
     if (this != &x) {
-        if (auto manager = std::exchange(_manager, x._manager)) {
```
compaction_backlog_tracker::compaction_backlog_tracker(compaction_backlog_tracker&& other) should be fixed the same way. if other is registered, we should allow it to be deregistered naturally via its ctor.  as in move assignment, the problem is benign in practice

bhalevy replied:
What do you mean?
This hunk deregisters `this` from `this->_manager`, but in the move ctor `this->_manager` will always be null.

bhalevy replied:
I opened https://github.com/scylladb/scylladb/issues/15248
I include a fix for it in the next revision

raphaelsc commented on replica/table.cc:
```diff
@@ -2044,20 +2044,42 @@ future<db::replay_position> table::discard_sstables(db_clock::time_point truncat
+        pruner& p;
+        db_clock::time_point truncated_at;
+        std::unordered_map<compaction_group*, compaction_backlog_tracker> new_trackers;
+
```
it seems you forgot to update backlog trackers here instead of post cache update when deleting the dropped sstables

raphaelsc commented on replica/table.cc:
```diff
@@ -713,6 +709,10 @@ class table_sstables_adder : public row_cache::external_updater_impl {
+        if (_is_main) {
+            _new_backlog_tracker.emplace(_cg.get_backlog_tracker().clone());
+            _new_backlog_tracker->replace_sstables({}, _sstables);
+        }
```
as we discussed, this is doing cloning work twice, as replace_sstables also clones the tracker state in order to provide strong exception guarantees. let's avoid redundant work by unifying clone and replace_sstables (the former is already done today anyway)

bhalevy replied:
next version will introduce clone_and_replace_sstables that will combine the exception-safe replace_sstables with cloning the tracker contents.

raphaelsc commented on sstables/sstable_set.cc:
```diff
@@ -1180,18 +1180,19 @@ compound_sstable_set::bytes_on_disk() const noexcept {
-        return boost::copy_range<std::vector<sstable_set::incremental_selector>>(_sets | boost::adaptors::transformed([] (const auto& set) {
+        // make_incremental_selector clones the underlying sstable_set impl
+        // so we essentially take a snapshot of the compound set contents here.
+        return boost::copy_range<std::vector<sstable_set::incremental_selector>>(sets | boost::adaptors::transformed([] (const auto& set) {
```
it doesn't really clone the set, does it? it seems to me make_incremental_selector() only takes a reference of shared ptr, assuming the underlying content is immutable.

raphaelsc replied:
but I don't think it should. otherwise reads will pay the price.

raphaelsc replied:
Ah I understand it now. Now we have each set managed by incremental selector taking a copy of _impl, to avoid that problem of range scan referencing outdated set.

raphaelsc replied:
But once we have compound sets managed under another compound set (thinking forward to tablets, I elabore further on https://github.com/scylladb/scylladb/pull/14043/commits/9977ccb917ff0c34787a95ae7758393f80574a1b#r1313937913), then taking a copy of _impl is not enough, as compound set may still be pointing to stable sstable_set addresses which may change during a range scan.

raphaelsc commented on sstables/sstable_set.cc:
```diff
@@ -342,6 +347,30 @@ seastar::shared_ptr<sstable_set_impl> partitioned_sstable_set::clone() const {
+    }
+
+    co_return seastar::make_shared<partitioned_sstable_set>(_schema, std::move(unleveled_sstables), std::move(leveled_sstables), std::move(all), std::move(all_runs), _use_level_metadata, _bytes_on_disk);
+}
```
partitioned_sstable_set needs another ctor. the existing one takes leveled_sstables, etc, as const ref, meaning copy won't be avoided.

raphaelsc commented on sstables/sstable_set.cc:
```diff
@@ -1118,18 +1118,25 @@ time_series_sstable_set::create_single_key_sstable_reader(
+    std::vector<sstable_set> cloned_sets = boost::copy_range<std::vector<sstable_set>>(_sets | boost::adaptors::transformed([] (const sstable_set* sst_set_ptr) {
+        return sst_set_ptr->clone();
+    }));
+    return seastar::make_shared<compound_sstable_set>(compound_sstable_set(_schema, std::move(cloned_sets)));
```
Does clone of compound set need really to invoke underlying sstable set's clone?

If we say that sstable_set_impl is still immutable, then compound set's clone only need to create a new sstable set with a copy of _impl (not clone full content), as we do for range scans now.

raphaelsc replied:
Maybe it's safer this way as compound set could be managing compound sets, and we want it work with any depth of compound sets.

raphaelsc commented on replica/table.cc:
```diff
@@ -136,10 +139,6 @@ sstables::sstable_set table::make_maintenance_sstable_set() const {
-void table::refresh_compound_sstable_set() {
-    _sstables = make_compound_sstable_set();
-}
-
```
if we want to avoid full refresh when a single sstable set changes, then compound set pointing to stable addresses is indeed a viable approach.

this approach is good in that compaction group no longer needs to be aware of compound set, and updates to each group are pulled automatically. 

if we think forward to tablets, the compound set for it will have to be aware of each compaciton_group, not individual sets as we want range scans to open a single group at a time, and filter out unwanted groups for single-key reads. See https://github.com/scylladb/scylladb/pull/13666/

table::make_compound_sstable_set() will have to be changed to produce one compound set for each group. so a tablet aware set will be managing a set of compound sets and their respective group ptrs (to collect metadata from).

avikivity commented:
Please make the subject more specific, it can lead people to think there is a data corruption problem.

tgrabiec commented on compaction/compaction.cc:
```diff
@@ -433,7 +433,7 @@ class compaction {
     std::vector<shared_sstable> _all_new_sstables;
-    lw_shared_ptr<sstable_set> _compacting;
+    optimized_optional<sstable_set> _compacting;
     const sstables::compaction_type _type;
```
I think the patch title "sstable_set: allow updating in-place" is misleading. We're laying the ground for that, but this patch is changing sstable_set to be a shared pointer instead of value object.

I think we don't want to update sstable_set in place, because it should remain a snapshot.

bhalevy replied:
Doing so creates complexity in taking a snapshot of the sstable_set in the for of a `lw_shared_ptr<sstable_set>` that we just got used to, but I don't think it's any less complicated than pushing the shared ptr one layer down, into the sstable_set itself, and having the sstable_set_impl contain the value object.

Plus, it forces us to recalculate the table's compound sstable_set when updating even a single sstable_set, the way compound_sstable_set is built today.

To prevent the latter, I think that we should move towards a state-less compound_sstable_set that will walk over the table's sstable_set:s dynamically.  FWIW, its incremental selector can take a snapshot of the incremental selectors of all sets when built, very much like it does today (and maybe even that is not required as you pointed out in https://github.com/scylladb/scylladb/pull/14043#discussion_r1315195879).

tgrabiec replied:
> pushing the shared ptr one layer down

This is fine

> it forces us to recalculate the table's compound sstable_set when updating even a single sstable_set, the way compound_sstable_set is built today.

Maybe we shouldn't keep per-table sstable set all the time and build it on demand instead?

Changing sstable_set underneath our feet violates immutability which causes correctness issues.

bhalevy replied:
A copy of an sstable_set is immutable, since only the original object will change in place. Very much like a copy of a `lw_shared_ptr<sstable_set>` that we keep today.

tgrabiec replied:
> A copy of an sstable_set is immutable, since only the original object will change in place. Very much like a copy of a lw_shared_ptr<sstable_set> that we keep today.

Just after this change, yes. But not when you make compound_sstable_set state-less. After that, copying table::_sstables is no longer taking a snapshot. That's why you had to clone the sstable_set deeply in table::sstables_as_snapshot_source(), to restore snapshot semantics. In every place which currently assumes snapshot semantics from lw_shared_ptr<sstable_set>, we would have to do that. How do we know it's the only place?

tgrabiec commented on replica/table.cc:
```diff
@@ -115,17 +115,20 @@ table::make_sstable_reader(schema_ptr s,
 
 sstables::sstable_set compaction_group::make_compound_sstable_set() {
-    return sstables::make_compound_sstable_set(_t.schema(), { _main_sstables, _maintenance_sstables });
+    return sstables::make_compound_sstable_set(_t.schema(), { &_main_sstables, &_maintenance_sstables });
```
Don't we lose snapshot semantics with this?

This can cause compound_sstable_set::for_each_sstable_gently_until() to skip an sstable when it moves from the maintenance set to the main set, and we already walked over the main set.

bhalevy replied:
for_each_sstable_gently_until takes a snapshot of all sstable_set_impl:s and walks over the snapshot, exactly because of this reason.

tgrabiec replied:
Right, this one is covered.

bhalevy commented on row_cache.hh:
```diff
@@ -179,6 +179,8 @@ public:
         // FIXME: make execute() noexcept, that will require every updater to make execution exception safe,
         // also change function signature.
+        // Updaters that are exception safe should handle all exceptions in prepare.
+        // Any exceptions from execute() are fatal.
```
So what should we do if rollback in execute fails?

tgrabiec commented on replica/table.cc:
```diff
@@ -1726,23 +1719,23 @@ table::make_partition_presence_checker(const sstables::sstable_set& ssts) {
 table::sstables_as_snapshot_source() {
     return snapshot_source([this] () {
-        auto sst_set = _sstables;
+        auto sst_set = make_lw_shared<sstable_set>(_sstables.clone());
```
We moved refresh_compound_sstable_set() here, which is still not exception safe, because in row cache we need to take the snapshot atomically with execute():

```
    [&] () noexcept {
        _prev_snapshot_pos = dht::ring_position::min();
        _prev_snapshot = std::exchange(_underlying, _snapshot_source());
```

bhalevy commented:
Submitted https://github.com/scylladb/scylladb/pull/15544

bhalevy commented:
In v8 ([bee3d52](https://github.com/scylladb/scylladb/commit/bee3d52cb89f61d22cecdacefea07deb5bf5fd9a)):
- reorganized series
- compaction_backlog_tracker: add clone_and_replace_sstables
- dht: token: add next helper
- table: compaction_group_for_token: extend verfication ifndef SCYLLA_BUILD_MODE_RELEASE
- table: add table_sstable_set

raphaelsc commented:
@bhalevy FYI: table_sstable_set is needed to fix https://github.com/scylladb/scylladb/issues/16876. we now need to rebase table_sstable_set on top of storage_group, which in turn may store 1 or more compaction_group (needed for splitting). so table_sstable_set works with compound set of each storage_group.
